src/System.Diagnostics.Process/src/System/Diagnostics/ProcessWaitState.Unix.cs

```diff
@@ -0,0 +1,503 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Diagnostics
+{
+    // Overview
+    // --------
+    // We have a few constraints we're working under here:
+    // - waitpid is used on Unix to get the exit status (including exit code) of a child process, but the first call
+    //   to it after the child has completed will reap the child removing the chance of subsequent calls getting status.
+    // - The Process design allows for multiple indendent Process objects to be handed out, and each of those
+    //   objects may be used concurrently with each other, even if they refer to the same underlying process.
+    //   Same with ProcessWaitHandle objects.  This is based on the Windows design where anyone with a handle to the
+    //   process can retrieve completion information about that process.
+    // - There is no good Unix equivalent to a process handle nor to being able to asynchronously be notified
+    //   of a process' exit (without more intrusive mechanisms like ptrace), which means such support
+    //   needs to be layered on top of waitpid.
+    // 
+    // As a result, we have the following scheme:
+    // - We maintain a static/shared table that maps process ID to ProcessWaitState objects.
+    //   Access to this table requires taking a global lock, so we try to minimize the number of
+    //   times we need to access the table, primarily just the first time a Process object needs
+    //   access to process exit/wait information and subsequently when that Process object gets GC'd.
+    // - Each process holds a ProcessWaitState.Holder object; when that object is constructed,
+    //   it ensures there's an appropriate entry in the mapping table and increments that entry's ref count.
+    // - When a Process object is dropped and its ProcessWaitState.Holder is finalized, it'll
+    //   decrement the ref count, and when no more process objects exist for a particular process ID,
+    //   that entry in the table will be cleaned up.
+    // - This approach effectively allows for multiple independent Process objects for the same process ID to all
+    //   share the same ProcessWaitState.  And since they are sharing the same wait state object,
+    //   the wait state object uses its own lock to protect the per-process state.  This includes
+    //   caching exit / exit code / exit time information so that a Process object for a process that's already
+    //   had waitpid called for it can get at its exit information.
```

Comment:
So it seems like the one problem with this approach is that if a process exits, but there are outstanding wait handles handed out (and rooted, so they can't be GC'd), then a new process is created and the pid is recycled, new calls to get that process's wait state will get the old processe's WaitState.  Is that correct?

I'm not really sure what we can do here, but we might as well call it out and say we understand this problem and there isn't a mitigation?